### PR TITLE
[4.0] [plg_sampledata_multilang] Convert to prepared statement

### PR DESCRIPTION
**Pull request by HLeithner**
# Summary of Changes
Updated SQL queries to prepared statements and made some cleanups around the queries.

### Testing Instructions
Use the plugin in all ways you can think of.

Install multilang sampledata

### Expected result
Nothing changed.

### CODE REVIEW

File: plugins/sampledata/multilang/multilang.php [hunk 1]
```diff
@@ -487,15 +488,15 @@ public function onAjaxSampledataApplyStep8()
...
 
 		$query
-			->clear()
 			->update($db->quoteName('#__extensions'))
 			->set($db->quoteName('enabled') . ' = 1')
-			->where($db->quoteName('name') . ' = ' . $db->quote($pluginName))
-			->where($db->quoteName('type') . ' = ' . $db->quote('plugin'));
+			->where($db->quoteName('name') . ' = :pluginname')
+			->where($db->quoteName('type') . ' = ' . $db->quote('plugin'))
+			->bind(':pluginname',$pluginName);
```
Comment (Quy) [inline plugins/sampledata/multilang/multilang.php]:
```suggestion
			->bind(':pluginname', $pluginName);
```

File: plugins/sampledata/multilang/multilang.php [hunk 2]
```diff
@@ -937,13 +954,12 @@ private function addAssociations($groupedAssociations)
...
-		$db    = Factory::getDbo();
+		$db    = $this->db;
 		$query = $db->getQuery(true);
 
 		// Add Module in Module menus.
-		$query->clear()
-			->insert($db->quoteName('#__modules_menu'))
-			->columns(array($db->quoteName('moduleid'), $db->quoteName('menuid')))
+		$query->insert($db->quoteName('#__modules_menu'))
+			->columns([$db->quoteName('moduleid'), $db->quoteName('menuid')])
```
Comment (Quy) [inline plugins/sampledata/multilang/multilang.php]:
```suggestion
			->columns($db->quoteName(['moduleid', 'menuid']))
```

**Comment by brianteeman** [comment]
again - looks like two unrelated files here

**Comment by Quy** [comment]
1st attempt:
> There is an error in a sample data plugin. Response is invalid.

2nd attempt:
> Step 4: Failed creating the Main menu (en-GB).

Nothing in PHP error log.

**Comment by HLeithner** [comment]
This requires https://github.com/joomla-framework/database/pull/162 too... maybe this is the problem.

**Comment by HLeithner** [comment]
@Quy can you try it again with the database Pr applied because I can't find a reason why it shouldn't work. Did it work without the PR?.

**Comment by Quy** [comment]
Works without the PR.

How to apply the database PR?

Here is step 4 which corresponds with the code that uses `bindArray`.
> Step 4: Specific Main Menus have been created as well as associated 'List All Categories' menu items!

**Comment by HLeithner** [comment]
Apply this file https://github.com/joomla-framework/database/pull/162.diff to the libaries/vendor folder should do it

**Comment by infograf768** [comment]
@HLeithner
Using https://github.com/joomla-framework/database/pull/162.diff solves issues here (tested with mysqli)

**Comment by infograf768** [comment]
@Quy
I modified the .diff to be correctly applied locally.

[database_patch.txt](https://github.com/joomla/joomla-cms/files/3372137/database_patch.txt)

File: plugins/sampledata/multilang/multilang.php [hunk 3]
```diff
@@ -552,18 +554,21 @@ private function enablePlugin($pluginName)
...
-			->where($db->quoteName('module') . ' = ' . $db->quote('mod_menu'))
-			->where($db->quoteName('language') . ' = ' . $db->quote('*'))
-			->where($db->quoteName('client_id') . ' = ' . $db->quote('0'))
-			->where($db->quoteName('position') . ' = ' . $db->quote('sidebar-right'));
+			->where(
+				[
+					$db->quoteName('client_id') . ' = 0',
+					$db->quoteName('module') . ' = ' . $db->quote('mod_menu'),
+					$db->quoteName('language') . ' = ' . $db->quote('*'),
+					$db->quoteName('position') . ' = ' . $db->quote('sidebar-right')
```
Comment (SharkyKZ) [inline plugins/sampledata/multilang/multilang.php]:
```suggestion
					$db->quoteName('position') . ' = ' . $db->quote('sidebar-right'),
```

File: plugins/sampledata/multilang/multilang.php [hunk 4]
```diff
@@ -897,17 +902,29 @@ private function addBlogMenuItem($itemLanguage, $categoryId)
...
+				->insert($db->quoteName('#__associations'));
 
 			foreach ($associations as $language => $id)
 			{
-				$query->values(((int) $id) . ',' . $db->quote($context) . ',' . $db->quote($key));
+				$query->values(implode(',', $query->bindArray(
+					[
+						$id,
+						$context,
+						$key
```
Comment (SharkyKZ) [inline plugins/sampledata/multilang/multilang.php]:
```suggestion
						$key,
```

File: plugins/sampledata/multilang/multilang.php [hunk 5]
```diff
@@ -1249,16 +1265,21 @@ protected function getInstalledlangs($cms_client = 'administrator')
...
-			->where($db->quoteName('type') . ' = ' . $db->quote('language'))
-			->where($db->quoteName('state') . ' = 0')
-			->where($db->quoteName('enabled') . ' = 1')
-			->where($db->quoteName('client_id') . ' = ' . (int) $client_id);
+			->where(
+				[
+					$db->quoteName('type') . ' = ' . $db->quote('language'),
+					$db->quoteName('state') . ' = 0',
+					$db->quoteName('enabled') . ' = 1',
+					$db->quoteName('client_id') . ' = :clientid'
```
Comment (SharkyKZ) [inline plugins/sampledata/multilang/multilang.php]:
```suggestion
					$db->quoteName('client_id') . ' = :clientid',
```

File: plugins/sampledata/multilang/multilang.php [hunk 6]
```diff
@@ -1249,16 +1265,21 @@ protected function getInstalledlangs($cms_client = 'administrator')
...
-			->where($db->quoteName('client_id') . ' = ' . (int) $client_id);
+			->where(
+				[
+					$db->quoteName('type') . ' = ' . $db->quote('language'),
+					$db->quoteName('state') . ' = 0',
+					$db->quoteName('enabled') . ' = 1',
+					$db->quoteName('client_id') . ' = :clientid',
+				]
+			)
+			->bind(':clientid', $client_id);
```
Comment (Quy) [inline plugins/sampledata/multilang/multilang.php]:
```suggestion
			->bind(':clientid', $client_id, ParameterType::INTEGER);
```

**Comment by SharkyKZ** [comment]
I have tested this item :white_check_mark: successfully on 0b6957412d7170f631ebf83bf34bc8c3a507014a<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/25433">issues.joomla.org/tracker/joomla-cms/25433</a>.</sub>

**Comment by Quy** [comment]
I have tested this item :white_check_mark: successfully on 0b6957412d7170f631ebf83bf34bc8c3a507014a<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/25433">issues.joomla.org/tracker/joomla-cms/25433</a>.</sub>

**Comment by Quy** [comment]
RTC<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/25433">issues.joomla.org/tracker/joomla-cms/25433</a>.</sub>

**Comment by zero-24** [comment]
```
FILE: /********/src/plugins/sampledata/multilang/multilang.php
38 | ----------------------------------------------------------------------
39 | FOUND 3 ERRORS AFFECTING 2 LINES
40 | ----------------------------------------------------------------------
41 | 926 \| ERROR \| [x] Multi-line function call not indented correctly;
42 | \| \| expected 20 spaces but found 16. Note: the
43 | \| \| autofixer will convert spaces to tabs
44 | \| \| (PEAR.Functions.FunctionCallSignature.Indent)
45 | 926 \| ERROR \| [x] Multi-line function call not indented correctly;
46 | \| \| expected 20 spaces but found 16. Note: the
47 | \| \| autofixer will convert spaces to tabs
48 | \| \| (PEAR.Functions.FunctionCallSignature.Indent)
49 | 927 \| ERROR \| [x] Closing parenthesis of a multi-line function call
50 | \| \| must be on a line by itself
51 | \| \| (PEAR.Functions.FunctionCallSignature.CloseBracketLine)
52 | ----------------------------------------------------------------------
53 | PHPCBF CAN FIX THE 3 MARKED SNIFF VIOLATIONS AUTOMATICALLY
54 | ----------------------------------------------------------------------
55


```

File: plugins/sampledata/multilang/multilang.php [hunk 7]
```diff
@@ -937,13 +957,12 @@ private function addAssociations($groupedAssociations)
...
+		$db    = $this->db;
 		$query = $db->getQuery(true);
 
 		// Add Module in Module menus.
-		$query->clear()
-			->insert($db->quoteName('#__modules_menu'))
-			->columns(array($db->quoteName('moduleid'), $db->quoteName('menuid')))
+		$query->insert($db->quoteName('#__modules_menu'))
+			->columns($db->quoteName(['moduleid', 'menuid']))
 			->values($moduleId . ', 0');
```
Comment (alikon) [inline plugins/sampledata/multilang/multilang.php]: shouldn't `$moduleId` be bounded as well ?

**Review by alikon — CHANGES_REQUESTED** [review]

File: plugins/sampledata/multilang/multilang.php (hunk 7 above)
Comment (HLeithner) [reply plugins/sampledata/multilang/multilang.php]: true, thx

**Comment by HLeithner** [comment]
We need a new test here please.

**Comment by Quy** [comment]
I have tested this item :white_check_mark: successfully on f1c7ccd3656d98015d6bd9e548196aff3f34328f<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/25433">issues.joomla.org/tracker/joomla-cms/25433</a>.</sub>

**Pull request merged**